laplace/baselaplace.py

```diff
@@ -1568,8 +1567,1025 @@ def sample(self, n_samples=100, generator=None):
+        self.seed = seed
+
+        self.K_MM = None
+        self.Sigma_inv = None  # (K_{MM} + L_MM_inv)^{-1}
+        self.train_loader = (
+            None  # needed in functional variance and marginal log likelihood
+        )
+        self.batch_size = None
+        self._prior_factor_sod = None
+        self.mu = None  # mean in the scatter term of the log marginal likelihood
+        self.L = None
+
+        # MAP estimate of NN parameters (used in regression marginal likelihood)
+        self.map_estimate = parameters_to_vector(self.model.parameters()).detach()
+
+        self._fitted = False
+        self._recompute_Sigma = True
+
+    @staticmethod
+    def _check_prior_precision(prior_precision):
+        """
+        Checks if the given prior precision is suitable for the GP interpretation of LLA.
+        As such, only single value priors, i.e., isotropic priors are suitable.
+        """
+        if torch.is_tensor(prior_precision):
+            if not (
+                prior_precision.ndim == 0
+                or (prior_precision.ndim == 1 and len(prior_precision) == 1)
+            ):
+                raise ValueError('Only isotropic priors supported in FunctionalLaplace')
+
+    def _init_K_MM(self):
+        """
+        Allocates memory for the kernel matrix evaluated at the subset of the training
+        data points. If the subset is of size \(M\) and the problem has \(C\) outputs,
+        this is a list of C \((M,M\)) tensors for diagonal kernel and \((M x C, M x C)\)
+        otherwise.
+        """
+        if self.diagonal_kernel:
+            self.K_MM = [
+                torch.empty(size=(self.M, self.M), device=self._device)
+                for _ in range(self.n_outputs)
+            ]
+        else:
+            self.K_MM = torch.empty(
+                size=(self.M * self.n_outputs, self.M * self.n_outputs),
+                device=self._device,
+            )
+
+    def _init_Sigma_inv(self):
+        """
+        Allocates memory for the cholesky decomposition of
+        \[
+            K_{MM} + \Lambda_{MM}^{-1}.
+        \]
+        See See [Improving predictions of Bayesian neural nets via local linearization (Immer et al., 2021)](https://arxiv.org/abs/2008.08400)
+        Equation 15 for more information.
+        """
+        if self.diagonal_kernel:
+            self.Sigma_inv = [
+                torch.empty(size=(self.M, self.M), device=self._device)
+                for _ in range(self.n_outputs)
+            ]
+        else:
+            self.Sigma_inv = torch.empty(
+                size=(self.M * self.n_outputs, self.M * self.n_outputs),
+                device=self._device,
+            )
+
+    def _store_K_batch(self, K_batch, i, j):
+        """
+        Given the kernel matrix between the i-th and the j-th batch, stores it in the
+        corresponding position in self.K_MM.
+        """
+        if self.diagonal_kernel:
+            for c in range(self.n_outputs):
```

Comment:
Is it possible to avoid the loop?

Reply:
I believe these loops where added by Metod to solve (or mitigate) memory problems in the GP approach. See [commit in PR](https://github.com/metodj/Laplace/pull/2/commits/ac74a0b8b6517b297d73ea12e4285d33461a99f5). I could revert them but those problems would (probably) raise again.

Reply:
Ok, makes sense 👌